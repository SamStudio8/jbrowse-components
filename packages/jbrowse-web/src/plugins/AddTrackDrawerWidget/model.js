import { types } from 'mobx-state-tree'
import { ElementId } from '../../mst-types'

export default types.model('AddTrackModel', {
  id: ElementId,
  type: types.literal('AddTrackDrawerWidget'),
})
