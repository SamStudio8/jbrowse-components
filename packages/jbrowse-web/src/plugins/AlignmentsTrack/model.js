import {
  types,
  getSnapshot,
  getParent,
  getRoot,
  flow,
  isAlive,
  setLivelinessChecking,
} from 'mobx-state-tree'

import { autorun } from 'mobx'
import { ConfigurationReference, getConf } from '../../configuration'

import { Region } from '../../mst-types'
import { renderRegionWithWorker } from '../../render'

import { BaseTrack as LinearGenomeTrack } from '../LinearGenomeView/models/model'
import AlignmentsTrack, {
  AlignmentsTrackBlock,
} from './components/AlignmentsTrack'

setLivelinessChecking('error')

const BlockState = types
  .model('BlockState', {
    key: types.string,
    region: Region,
  })
  .volatile(() => ({
    filled: false,
    data: undefined,
    reactComponent: AlignmentsTrackBlock,
    html: '',
    error: undefined,
  }))
  .views(self => ({
    get rendererType() {
      const { pluginManager } = getRoot(self)
      const track = getParent(self, 2)
      const RendererType = pluginManager.getRendererType(track.rendererType)
      if (!RendererType)
        throw new Error(`renderer "${track.rendererType} not found`)
      if (!RendererType.ReactComponent)
        throw new Error(
          `renderer ${
            track.rendererType
          } has no ReactComponent, it may not be completely implemented yet`,
        )
      return RendererType
    },

    get renderProps() {
      const view = getParent(self, 4) // view -> [tracks] -> [blocks]
      return { bpPerPx: view.bpPerPx }
    },
  }))
  .actions(self => ({
    afterAttach() {
      self.render()
    },
    render: flow(function* renderBlock() {
      if (!isAlive(self)) return
      const track = getParent(self, 2)
      const view = getParent(track, 2)
      const root = getParent(view, 2)
      try {
        // console.log('calling', self.region.toJSON())
        const { html, ...data } = yield renderRegionWithWorker(root.app, {
          region: self.region,
          adapterType: track.adapterType.name,
          adapterConfig: getConf(track, 'adapter'),
          rendererType: track.rendererType,
          renderProps: self.renderProps,
          sessionId: track.id,
          timeout: 10000,
        })
        if (!isAlive(self)) return
        self.filled = true
        self.data = data
        self.html = html
        // console.log('finished', self.region.toJSON(), self.html)
      } catch (error) {
        // the rendering failed for some reason
        console.error(error)
        self.error = error
      }
    }),
  }))

const BlockBasedTrackState = types.compose(
  'BlockBasedTrackState',
  LinearGenomeTrack,
  types
    .model({
      blockState: types.map(BlockState),
    })
    .actions(self => {
      let blockWatchDisposer
      function disposeBlockWatch() {
        if (blockWatchDisposer) blockWatchDisposer()
        blockWatchDisposer = undefined
      }
      return {
        afterAttach() {
          const view = getParent(self, 2)
          // watch the parent's blocks to update our block state when they change
          blockWatchDisposer = autorun(() => {
            // create any blocks that we need to create
            const blocksPresent = {}
            view.blocks.forEach(block => {
              blocksPresent[block.key] = true
              if (!self.blockState.has(block.key))
                self.blockState.set(
                  block.key,
                  BlockState.create({
                    key: block.key,
                    region: block,
                  }),
                )
            })
            // delete any blocks we need to delete
            self.blockState.forEach((value, key) => {
              if (!blocksPresent[key]) self.blockState.delete(key)
            })
          })
        },
        beforeDetach: disposeBlockWatch,
        beforeDestroy: disposeBlockWatch,
      }
    }),
)

export default (pluginManager, configSchema) =>
  types.compose(
    'AlignmentsTrack',
    BlockBasedTrackState,
    types
      .model({
        type: types.literal('AlignmentsTrack'),
        configuration: ConfigurationReference(configSchema),
        // the renderer that the user has selected in the UI, empty string
        // if they have not made any selection
        selectedView: types.optional(types.string, ''),
        height: types.optional(types.integer, 100),
      })
      .volatile(self => ({
        reactComponent: AlignmentsTrack,
      }))
      // .actions(self => ({
      //   afterAttach() {
      //     onPatch(self, patch => {
      //       console.log('patch', self.name, patch)
      //     })
      //   },
      // }))
      .views(self => ({
        // the renderer type is based on the "view" selected in the UI: pileup, coverage, etc
        get rendererType() {
          const defaultView = getConf(self, 'defaultView')
          const viewName = self.selectedView || defaultView
          const rendererType = { pileup: 'PileupRenderer' }[viewName]
          if (!rendererType)
            throw new Error(`unknown alignments view name ${viewName}`)
          return rendererType
        },

        get adapterType() {
          const adapterConfig = getConf(self, 'adapter')
          if (!adapterConfig)
            throw new Error(
              `no adapter configuration provided for ${self.type}`,
            )
          const adapterType = pluginManager.getAdapterType(adapterConfig.type)
          if (!adapterType)
            throw new Error(`unknown adapter type ${adapterConfig.type}`)
          return adapterType
        },

        get adapter() {
          // TODO: refactor this someplace else to make this code shareable
          const adapter = new self.adapterType.AdapterClass(
            getSnapshot(self.configuration.adapter),
          )
          return adapter
        },
      })),
  )
