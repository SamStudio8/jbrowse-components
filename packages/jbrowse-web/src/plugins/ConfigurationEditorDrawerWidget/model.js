import { types } from 'mobx-state-tree'
import { ElementId } from '../../mst-types'
import { Assembly } from '../../rootModel'

export default pluginManager =>
  types
    .model('ConfigurationEditorDrawerWidget', {
      id: ElementId,
      type: types.literal('ConfigurationEditorDrawerWidget'),
      target: types.reference(
        types.union(
          pluginManager.pluggableConfigSchemaType('track'),
          pluginManager.pluggableConfigSchemaType('view'),
          Assembly,
        ),
      ),
    })
    // .volatile(() => ({
    //   target: undefined,
    // }))
    .actions(self => ({
      setTarget(newTarget) {
        self.target = newTarget
      },
    }))
//    .views(self => ({}))
