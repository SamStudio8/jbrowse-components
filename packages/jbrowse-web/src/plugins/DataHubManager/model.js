import { types } from 'mobx-state-tree'
import { ElementId } from '../../mst-types'

export default types.model('DataHubModel', {
  id: ElementId,
  type: types.literal('DataHubDrawerWidget'),
})
