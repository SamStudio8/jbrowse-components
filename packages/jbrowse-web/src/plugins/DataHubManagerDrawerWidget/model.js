import { types } from 'mobx-state-tree'
import { ElementId } from '@gmod/jbrowse-core/mst-types'

export default types.model('DataHubModel', {
  id: ElementId,
  type: types.literal('DataHubDrawerWidget'),
})
