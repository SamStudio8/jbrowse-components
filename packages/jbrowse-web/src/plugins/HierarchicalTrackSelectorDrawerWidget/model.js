import { types, onPatch, getRoot } from 'mobx-state-tree'
import { readConfObject, getConf } from '../../configuration'
import { ElementId } from '../../mst-types'

export function generateHierarchy(trackConfigurations) {
  const hierarchy = {}

  trackConfigurations.forEach(trackConf => {
    const categories = [...(readConfObject(trackConf, 'category') || [])]
    if (categories.length === 0) categories[0] = 'uncategorized'

    let currLevel = hierarchy
    let i = 0
    for (; i < categories.length; i += 1) {
      const category = categories[i]
      if (!currLevel[category]) currLevel[category] = {}
      currLevel = currLevel[category]
    }
    currLevel[
      readConfObject(trackConf, 'name') || trackConf._configId
    ] = trackConf
  })
  return hierarchy
}

export default pluginManager =>
  types.compose(
    'HierarchicalTrackSelectorDrawerWidget',
    types
      .model({
        id: ElementId,
        type: types.literal('HierarchicalTrackSelectorDrawerWidget'),
        collapsed: types.map(types.boolean), // map of category path -> boolean of whether it is collapsed
        filterText: '',
        view: types.reference(
          pluginManager.pluggableMstType('view', 'stateModel'),
        ),
      })
      .actions(self => ({
        afterAttach() {
          onPatch(self, patch => console.log(patch))
        },
        setView(view) {
          self.view = view
        },
        toggleCategory(pathName) {
          self.collapsed.set(pathName, !self.collapsed.get(pathName))
        },
        clearFilterText() {
          self.filterText = ''
        },
        setFilterText(newText) {
          self.filterText = newText
        },
      }))
      .views(self => ({
        get trackConfigurations() {
          const root = getRoot(self)
          const trackConfigurations = root.configuration.tracks
          const relevantTrackConfigurations = trackConfigurations.filter(
            conf => conf.viewType === self.view.type,
          )
          return relevantTrackConfigurations
        },

        get hierarchy() {
          return generateHierarchy(self.trackConfigurations)
        },
      })),
  )
