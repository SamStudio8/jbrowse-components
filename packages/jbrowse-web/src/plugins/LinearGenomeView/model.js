import shortid from 'shortid'
import { types, getParent, getSnapshot } from 'mobx-state-tree'
import { ConfigurationSchema } from '../../util/configuration'

const IdType = types.optional(types.identifier, shortid.generate)

export const Block = types
  .model({
    refName: types.string,
    start: types.number,
    end: types.number,
  })
  .views(self => ({
    get widthPx() {
      return Math.abs(self.end - self.start) / getParent(self, 2).bpPerPx
    },
  }))

const minTrackHeight = 20
export const Track = types.model('Track', {
  id: IdType,
  name: types.string,
  type: types.string,
  height: types.optional(
    types.refinement('trackHeight', types.number, n => n >= minTrackHeight),
    minTrackHeight,
  ),
  subtracks: types.maybe(types.array(types.late(() => Track))),
  configuration: ConfigurationSchema('Track', {
    backgroundColor: {
      description: `the track's background color`,
      type: 'color',
      defaultValue: '#eee',
    },
  }),
})

const ViewStateBase = types.model({
  // views have an auto-generated ID by default
  id: IdType,
})

const LinearGenomeViewState = types
  .compose(
    'LinearViewModel',
    ViewStateBase,
    types.model({
      type: types.literal('linear'),
      offsetPx: 0,
      bpPerPx: 1,
      blocks: types.array(Block),
      tracks: types.array(Track),
      controlsWidth: 100,
      width: 800,
      configuration: ConfigurationSchema('LinearView', {
        bar: { type: 'integer', model: types.integer, defaultValue: 0 },
      }),
    }),
  )
  .views(self => ({
    get totalBlocksWidthPx() {
      return self.blocks.reduce((a, b) => a + b.widthPx, 0)
    },
  }))
  .actions(self => ({
    addTrack(id, name, type) {
      self.tracks.push(Track.create({ id, name, type }))
    },

    pushBlock(refName, start, end) {
      self.blocks.push(Block.create({ refName, start, end }))
    },

    horizontalScroll(distance) {
      self.offsetPx = Math.min(
        Math.max(self.offsetPx + distance, 0),
        self.width - self.totalBlocksWidthPx - self.controlsWidth,
      )
    },
  }))

export default LinearGenomeViewState
