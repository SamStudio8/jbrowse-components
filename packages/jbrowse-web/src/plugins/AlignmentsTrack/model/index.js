import { types } from 'mobx-state-tree'
import { ElementId } from '../../../mst-types'
import AlignmentsTrack from './alignmentsTrack'

export default AlignmentsTrack

export const AlignmentsFeatureDrawerWidgetModel = types
  .model('AlignmentsFeatureDrawerWidget', {
    id: ElementId,
    type: types.literal('AlignmentsFeatureDrawerWidget'),
  })
  .volatile(() => ({
    featureData: {},
  }))
  .actions(self => ({
    setFeatureData(data) {
      self.featureData = data
    },
    clearFeatureData() {
      self.featureData = {}
    },
  }))
