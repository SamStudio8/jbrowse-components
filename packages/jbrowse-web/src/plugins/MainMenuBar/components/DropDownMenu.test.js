import { createShallow } from '@material-ui/core/test-utils'
import React from 'react'
import { MainMenuBarModel } from '../model'
import DropDownMenu from './DropDownMenu'

describe('<DropDownMenu />', () => {
  let shallow

  beforeAll(() => {
    shallow = createShallow()
  })

  it('renders', () => {
    const menubar = MainMenuBarModel.create({
      id: 'testingId',
      type: 'MainMenuBar',
    })
    const menu = menubar.menus[0]
    const wrapper = shallow(
      <DropDownMenu
        model={menubar}
        menuTitle={menu.name}
        menuItems={menu.menuItems}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('opens and closes a menu and selects a menu item', () => {
    const menubar = MainMenuBarModel.create({
      id: 'testingId',
      type: 'MainMenuBar',
    })
    const menu = menubar.menus[0]
    const wrapper = shallow(
      <DropDownMenu
        model={menubar}
        menuTitle={menu.name}
        menuItems={menu.menuItems}
      />,
    )
      .first()
      .shallow()
    const instance = wrapper.instance()
    const helpButton = wrapper.find('WithStyles(Button)')
    // Click menu button to open
    instance.handleToggle({ currentTarget: helpButton })
    expect(wrapper).toMatchSnapshot()
    // Click menu button again to close (two events, one for menu click and one for ClickAwayListener)
    instance.handleClose({ target: 'Help' })
    expect(wrapper).toMatchSnapshot()
    instance.handleToggle({ currentTarget: helpButton })
    expect(wrapper).toMatchSnapshot()
    // Click menu button to open
    instance.handleToggle({ currentTarget: helpButton })
    expect(wrapper).toMatchSnapshot()
    // Click somewhere else on page to close with ClickAwayListener
    instance.handleClose({ target: 'somewhereElseOnPage' })
    expect(wrapper).toMatchSnapshot()
  })

  it('selects a menu item', () => {
    const menubar = MainMenuBarModel.create({
      id: 'testingId',
      type: 'MainMenuBar',
    })
    const menu = menubar.menus[0]
    const wrapper = shallow(
      <DropDownMenu
        model={menubar}
        menuTitle={menu.name}
        menuItems={menu.menuItems}
      />,
    )
      .first()
      .shallow()
    const instance = wrapper.instance()
    const helpButton = wrapper.find('WithStyles(Button)')
    const callback = jest.fn(() => {})
    // Click menu button to open
    instance.handleToggle({ currentTarget: helpButton })
    expect(wrapper).toMatchSnapshot()
    // Click menu item to close and activate callback
    instance.handleClose({ target: 'someMenuItem' }, callback)
    expect(wrapper).toMatchSnapshot()
    expect(callback.mock.calls.length).toBe(1)
  })
})
