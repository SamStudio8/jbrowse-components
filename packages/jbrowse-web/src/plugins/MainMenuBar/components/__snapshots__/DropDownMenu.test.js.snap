// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropDownMenu /> opens and closes a menu and selects a menu item 1`] = `
<div
  className="DropDownMenu-root-1"
>
  <WithStyles(Button)
    aria-haspopup="true"
    aria-owns="menu-list-grow"
    color="inherit"
    onClick={[Function]}
  >
    Help
    <WithStyles(Icon)>
      arrow_drop_down
    </WithStyles(Icon)>
  </WithStyles(Button)>
  <Popper
    anchorEl={
      <WithStyles(Button)
        aria-haspopup="true"
        aria-owns={null}
        color="inherit"
        onClick={[Function]}
      >
        Help
        <WithStyles(Icon)>
          arrow_drop_down
        </WithStyles(Icon)>
      </WithStyles(Button)>
    }
    className="DropDownMenu-popper-2"
    disablePortal={false}
    open={true}
    placement="bottom-start"
    transition={true}
  >
    <Component />
  </Popper>
</div>
`;

exports[`<DropDownMenu /> opens and closes a menu and selects a menu item 2`] = `
<div
  className="DropDownMenu-root-1"
>
  <WithStyles(Button)
    aria-haspopup="true"
    aria-owns="menu-list-grow"
    color="inherit"
    onClick={[Function]}
  >
    Help
    <WithStyles(Icon)>
      arrow_drop_down
    </WithStyles(Icon)>
  </WithStyles(Button)>
  <Popper
    anchorEl={
      <WithStyles(Button)
        aria-haspopup="true"
        aria-owns={null}
        color="inherit"
        onClick={[Function]}
      >
        Help
        <WithStyles(Icon)>
          arrow_drop_down
        </WithStyles(Icon)>
      </WithStyles(Button)>
    }
    className="DropDownMenu-popper-2"
    disablePortal={false}
    open={true}
    placement="bottom-start"
    transition={true}
  >
    <Component />
  </Popper>
</div>
`;

exports[`<DropDownMenu /> opens and closes a menu and selects a menu item 3`] = `
<div
  className="DropDownMenu-root-1"
>
  <WithStyles(Button)
    aria-haspopup="true"
    aria-owns={null}
    color="inherit"
    onClick={[Function]}
  >
    Help
    <WithStyles(Icon)>
      arrow_drop_down
    </WithStyles(Icon)>
  </WithStyles(Button)>
  <Popper
    anchorEl={null}
    className="DropDownMenu-popper-2"
    disablePortal={false}
    open={false}
    placement="bottom-start"
    transition={true}
  >
    <Component />
  </Popper>
</div>
`;

exports[`<DropDownMenu /> opens and closes a menu and selects a menu item 4`] = `
<div
  className="DropDownMenu-root-1"
>
  <WithStyles(Button)
    aria-haspopup="true"
    aria-owns="menu-list-grow"
    color="inherit"
    onClick={[Function]}
  >
    Help
    <WithStyles(Icon)>
      arrow_drop_down
    </WithStyles(Icon)>
  </WithStyles(Button)>
  <Popper
    anchorEl={
      <WithStyles(Button)
        aria-haspopup="true"
        aria-owns={null}
        color="inherit"
        onClick={[Function]}
      >
        Help
        <WithStyles(Icon)>
          arrow_drop_down
        </WithStyles(Icon)>
      </WithStyles(Button)>
    }
    className="DropDownMenu-popper-2"
    disablePortal={false}
    open={true}
    placement="bottom-start"
    transition={true}
  >
    <Component />
  </Popper>
</div>
`;

exports[`<DropDownMenu /> opens and closes a menu and selects a menu item 5`] = `
<div
  className="DropDownMenu-root-1"
>
  <WithStyles(Button)
    aria-haspopup="true"
    aria-owns={null}
    color="inherit"
    onClick={[Function]}
  >
    Help
    <WithStyles(Icon)>
      arrow_drop_down
    </WithStyles(Icon)>
  </WithStyles(Button)>
  <Popper
    anchorEl={null}
    className="DropDownMenu-popper-2"
    disablePortal={false}
    open={false}
    placement="bottom-start"
    transition={true}
  >
    <Component />
  </Popper>
</div>
`;

exports[`<DropDownMenu /> renders 1`] = `
<Component
  classes={
    Object {
      "grow": "DropDownMenu-grow-3",
      "popper": "DropDownMenu-popper-2",
      "root": "DropDownMenu-root-1",
    }
  }
  menuItems={
    Array [
      Object {
        "callback": "openAbout",
        "icon": "info",
        "name": "About",
      },
      Object {
        "callback": "openHelp",
        "icon": "help",
        "name": "Help",
      },
      Object {
        "callback": "openAssemblyEditor",
        "icon": "edit",
        "name": "Assemblies",
      },
    ]
  }
  menuTitle="Help"
  model={
    Object {
      "id": "testingId",
      "menus": Array [
        Object {
          "menuItems": Array [
            Object {
              "callback": "openAbout",
              "icon": "info",
              "name": "About",
            },
            Object {
              "callback": "openHelp",
              "icon": "help",
              "name": "Help",
            },
            Object {
              "callback": "openAssemblyEditor",
              "icon": "edit",
              "name": "Assemblies",
            },
          ],
          "name": "Help",
        },
      ],
      "type": "MainMenuBar",
    }
  }
/>
`;

exports[`<DropDownMenu /> selects a menu item 1`] = `
<div
  className="DropDownMenu-root-1"
>
  <WithStyles(Button)
    aria-haspopup="true"
    aria-owns="menu-list-grow"
    color="inherit"
    onClick={[Function]}
  >
    Help
    <WithStyles(Icon)>
      arrow_drop_down
    </WithStyles(Icon)>
  </WithStyles(Button)>
  <Popper
    anchorEl={
      <WithStyles(Button)
        aria-haspopup="true"
        aria-owns={null}
        color="inherit"
        onClick={[Function]}
      >
        Help
        <WithStyles(Icon)>
          arrow_drop_down
        </WithStyles(Icon)>
      </WithStyles(Button)>
    }
    className="DropDownMenu-popper-2"
    disablePortal={false}
    open={true}
    placement="bottom-start"
    transition={true}
  >
    <Component />
  </Popper>
</div>
`;

exports[`<DropDownMenu /> selects a menu item 2`] = `
<div
  className="DropDownMenu-root-1"
>
  <WithStyles(Button)
    aria-haspopup="true"
    aria-owns={null}
    color="inherit"
    onClick={[Function]}
  >
    Help
    <WithStyles(Icon)>
      arrow_drop_down
    </WithStyles(Icon)>
  </WithStyles(Button)>
  <Popper
    anchorEl={null}
    className="DropDownMenu-popper-2"
    disablePortal={false}
    open={false}
    placement="bottom-start"
    transition={true}
  >
    <Component />
  </Popper>
</div>
`;
